frontend/persist.lua

```diff
@@ -0,0 +1,133 @@
+            end
+            return str
+        end,
+
+        deserialize = function(file)
+            local f, err, str
+            f, err = io.open(file, "rb")
+            if not f then
+                return nil, err
+            end
+            str, err = f:read("*a")
+            f:close()
+            if not str then
+                return nil, err
+            end
+            local ok, t = pcall(bitser.loads, str)
+            if not ok then
+                return nil, "malformed serialized data"
+            end
+            return t
+        end,
+    },
+    -- dump: human readable, pretty printed, fast enough for most user cases.
+    dump = {
+        id = "Dump serializer",
+
+        serialize = function(t, file, as_bytecode)
+            local content, err = dump(t)
+            if not content then
+                return nil, string.format("cannot serialize table %s: %s", t, err)
+            end
+            local str
+            if as_bytecode then
+                str, err =  load("return " .. content)
+                if not str then
+                    print("cannot convert table to bytecode: %s, ignoring", err)
+                else
+                    str = string.dump(str, true)
+                end
+            end
+
+            if not str then
+                str = "return " .. content
+            end
+            return str
+        end,
+
+        deserialize = function(file)
+            local ok, t, err = pcall(dofile, file)
+            if not ok then
+                return nil, err
+            end
+            return t
+        end,
+    }
+}
+
+local Persist = {}
+
+function Persist:new(o)
+    o = o or {}
+    setmetatable(o, self)
+    self.__index = self
+    return o:init(o.path, o.codec)
+end
+
+function Persist:init(path, codec)
+    if type(path) ~= "string" then
+        return nil, "path is required"
+    end
+    self.path = path
+    self.codec = codec or "dump"
+    return self
+end
+
+function Persist:stats(do_print)
+    local lfs = require("libs/libkoreader-lfs")
+    local size = lfs.attributes(self.path, "size") or -1
+    local timestamp = lfs.attributes(self.path, "modification")
+    local exists = size ~= -1
+
+    if do_print then
+        print(string.format("file: %s, codec: %s, exists: %s, size: %d, timestamp: %d",
+            self.path, self.codec, exists, size, timestamp))
+    else
+        return self.path, self.codec, exists, size, timestamp
+    end
+end
+
+function Persist:loadFile()
+    local t, err = codecs[self.codec]:deserialize(self.path)
+    if not t then
+        return nil, err
+    end
+    return t
+end
+
+function Persist:saveFile(t, as_bytecode)
+    local str, ok, err
+    str, err = codecs[self.codec].serialize(t, self.path, as_bytecode)
+    if not str then
+        return nil, err
+    end
+    ok, err = self:writeFile(str)
+    if not ok then
+        return nil, err
+    end
+end
+
+function Persist:writeFile(str)
```

Comment:
Should this one be public ?
If not, either rename it to `Persist:_writeFile()` so we know it is not - or inline it in `:saveFile()` )

Reply:
I think it is a legacy workaround while I played with bytecode. No longer needed. I'm going to inline it in saveFile